Create base interface for a RoboCommand Factory

RFBomb opened the pull request:
This PR adds a new object called RoboCommandFactory that contains methods for creating new robocommands. 

This object simply translates the constructors for `RoboCommand` into methods within an object. 
But the real reason for the PR is to introduce the `IRoboCommandFactoryBase` interface, which is meant to be a base interface for any objects that will be used to generate IRoboCommand objects. 

For example, in my dll that I am building, I require a specific set of constructors to generate my robocommands. These were custom-built constructors used for a class derived from RoboCommand. Since refactoring out of the main project into a dll, I now need to essentially allow the dll consumer to specify a factory object to create the robocommands. Only way to enforce the constructor requirements here is via an interface imposed on an object, then allow the consumer to design their own object that implements the interface. Since other consumers may one day need something similar, I introduced the factory here for their use.

RFBomb committed:
Create base interface for a RoboCommand Factory

RFBomb committed:
Fix Owner

RFBomb committed:
Rename Interface

PCAssistSoftware commented:
Hi @RFBomb, assuming this doesn't break anything in existing project then happy to merge this for you when you are ready.  Not sure how to test it.

As you have incremeted version number might just be worth changing copyright year in RoboSharp.csproj to 2022 at same time as well

RFBomb committed:
Update XML Comment

PCAssistSoftware commented:
@RFBomb - I assume the complete lack of response to my message means it is not ready to merge yet?

RFBomb commented:
@PCAssistSoftware 
- Sorry for delayed response (to your message that I can't seem to locate but have in an email from github?), this was just something that I whipped up while working on a dll that utilizes this dll. 

I will work on updating the unit tests for this to show how it works. But essentially its just a static object that contains two methods that merely point to the constructor for RoboCommand. 

The idea behind this is that if a consumer (like myself) wants to create a class that overrides the functionality of robocommand, and pass that factory to another consumer, that can now be facilitated. 

For example, I have a dll that will be creating robocopy commands to sync folders from one place to another. The consumer of that dll will be two seperate programs. One of them requires enabling logging, the other does not. Now, each of those 2 programs simply creates the factory object, passes it to the dll that generates the commands, and voila. The dll generated the commands according to the specifications of the factory methods defined by each consumer.

PCAssistSoftware commented:
Hi @RFBomb - that is odd, I had replied twice, but both messages gone for some reason??!

Let me know when ready to merge and I will get it done for you :)

RFBomb committed:
Force Pass error tests if running on AppVeyor`

OnError event is not thrown when running on AppVeyor for unknown reason. This check passes the test automatically if the test destination is exactly what appveyor uses.

RFBomb committed:
Add Enums for CopyOptions and SelectionOptions factory usage

RFBomb committed:
Update RoboCommand.cs

RFBomb committed:
Rename IRoboFactory to IRoboCommandFactory

RFBomb committed:
Rework RoboCommand Factory

RFBomb commented:
@PCAssistSoftware 
Check this last push out. 
This commit should hugely clarify what my design intent was for the factory. Essentially, a way for consumers to more easily create the robocommand (and any custom derived objects). 

So instead of using the constructors to produce the command, the factory can be called. The factory has been reworked from it (dumbly) just calling the constructor to now actually being an object to more easily create consumer factories.
 (Overriding the parameterless method to return a custom IRoboCommand instead of RoboCommand object affects the entire factory automatically). 

I have also introduced 2 new enums here. One resides within CopyOptions and the other within SelectionOptions for more compact construction of those objects (and thus eliminating potentially large amount of code thanks to also being able to use them via the factory methods)

RFBomb committed:
Update Test_Setup.cs

RFBomb committed:
Update SelectionOptions.cs

PCAssistSoftware commented:
That all make sense - I like what you have done here, will be very useful - great work (as always)

RFBomb committed:
Rename Factory Method

FromSourceAndDestination renamed to GetRoboCommand
All overloads now are named the same thing, because they fundamentally all do the same thing.

RFBomb committed:
Update RoboQueueResults.cs

RFBomb committed:
Update Remarks

RFBomb commented:
Ok, I believe this is ready for merge, unless you wanted to test it/have input on it. 

This is an example how I'm using it:
 - I have a 'RoboCopy' class that I created that is derived from RoboCommand. I override a few settings and enforce a couple defaults on it. 
 - I have a library I built that generates robocopy commands. 
 - To enforce the library to build using the customized object, I pass the new factory object (shown in picture below) to the library. Then, all the commands it generated will be of my customized type, instead of the base robocommand type.
![image](https://user-images.githubusercontent.com/20431767/168392126-0b164043-c86c-4c5d-b37c-cac848feeee3.png)

PCAssistSoftware commented:
All looks good to me, happy to go ahead and merge it for you now if you want?
